tempesta_fw/ss_skb.c

```diff
@@ -527,70 +575,102 @@ static int
-	       " nr_frags=%u frag_list=%p)...\n",
+	       " nr_frags=%u)...\n",
 	       len, pspt, skb->head, skb->data,
 	       skb_tail_pointer(skb), skb_end_pointer(skb),
 	       skb->len, skb->data_len, skb->truesize,
-	       skb_shinfo(skb)->nr_frags, skb_shinfo(skb)->frag_list);
+	       skb_shinfo(skb)->nr_frags);
+	BUG_ON(!len);
+
+	if ((len > (int)PAGE_SIZE) || (-len > (int)PAGE_SIZE)) {
+		SS_WARN("SKB room to add or delete is too big: %u\n", len);
+		return -EINVAL;
+	}
 
-	if (pspt >= vaddr && pspt < vaddr + dlen) {
+	/*
+	 * Determine where the split begins within the SKB,
+	 * then do the job using the right function.
+	 */
+
+	/* See if the split begins in the linear data. */
+	d_size = skb_headlen(skb);
+	offset = pspt - (char *)skb->data;
+
+	if ((offset >= 0) && (offset < d_size)) {
 		it->ptr = __split_linear_data(skb, pskb, pspt, len);
 		goto done;
 	}
 
+	/* See if the split begins in the page fragments data. */
 	for (i = 0; i < skb_shinfo(skb)->nr_frags; ++i) {
 		const skb_frag_t *frag = &skb_shinfo(skb)->frags[i];
-		vaddr = skb_frag_address(frag);
-		dlen = skb_frag_size(frag);
+		d_size = skb_frag_size(frag);
+		offset = pspt - (char *)skb_frag_address(frag);
 
-		if (pspt >= vaddr && pspt < vaddr + dlen) {
-			it->ptr = __split_pgfrag(skb, pskb, i, pspt - vaddr,
-						 len);
+		if ((offset >= 0) && (offset < d_size)) {
+			it->ptr = __split_pgfrag(skb, pskb, i, offset, len);
 			goto done;
 		}
 	}
 
-	skb_walk_frags(skb, frag_i) {
-		int r = __skb_fragment(frag_i, skb, pspt, len, it);
-		if (r != -ENOENT)
-			return r;
-	}
-
-	TFW_WARN("Trying to fragment skb by invalid pointer (%p, %p:%u)\n",
+	/* The split is not within the SKB. */
+	TFW_WARN("Attempt to fragment an SKB at wrong address (%p, %p:%u)\n",
 		 pspt, skb->head, skb->truesize);
 	return -ENOENT;
+
 done:
 	SS_DBG("%s: res=%p, skb: head=%p data=%p tail=%p end=%p"
-	       " len=%u data_len=%u truesize=%u"
-	       " nr_frags=%u frag_list=%p)\n", __func__, it->ptr,
-	       skb->head, skb->data,
+	       " len=%u data_len=%u truesize=%u nr_frags=%u\n",
+	       __func__, it->ptr, skb->head, skb->data,
 	       skb_tail_pointer(skb), skb_end_pointer(skb),
 	       skb->len, skb->data_len, skb->truesize,
-	       skb_shinfo(skb)->nr_frags, skb_shinfo(skb)->frag_list);
-
+	       skb_shinfo(skb)->nr_frags);
+	/*
+	 * Note: If data was deleted, then it's possible that @it->ptr
+	 * actually points at the address in a different (new) SKB.
+	 */
 	if (!it->ptr)
-		return SS_DROP;
+		return -EFAULT;
 	it->len = len;
 	it->skb = skb;
 	return 0;
 }
 
+/*
+ * Move an SKB from frag_list to the proper place in the list of SKB
+ * for an HTTP message.
+ *
+ * During the manipulations on SKB fragments a new SKB may be created
+ * to hold new or relocated fragments. The new SKB is stored in the
+ * original SKB's frag_list. After the manipulations on fragments are
+ * completed, the new SKB is inserted after the original SKB in the
+ * list of SKBs for an HTTP message.
```

Comment:
The comment is unclear. What is "proper place"? Currently there are strict order for skb processing, so fragments were also at proper place in sense of processing order. It's clear from the code that skb is moved from fragment to Tempesta's list of skbs. However, it's absolutely unclear why fragments don't work for us anymore and we need to linearize the skb list.